Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift

```diff
@@ -11,17 +11,116 @@
     // The currently active state.
-    private var currentState: VariableMap<Type> {
-        return stack.last!
+    private var currentState = VariableMap<Type>()
+
+    public mutating func reset() {
+        stateChangesStack = [[VariableMap<Type>()], [VariableMap<Type>()]]
+        currentState = VariableMap<Type>()
+    }
+
+    /// Update current variable types and type changes
+    /// Used after block end when some states should be merged to parent
+    public mutating func mergeStates(typeChanges: inout [(Variable, Type)]) {
+        let states = stateChangesStack.removeLast()
+        // Active state of the previous states entry
+        let prevState = stateChangesStack.last!.last!
+        var numUpdatesPerVariable = VariableMap<Int>()
+        var newTypes = VariableMap<Type>()
+
+        for state in states {
+            for (v, t) in state {
+                // Skip variable types that are out of scope as we do not care about them anymore
+                guard t != .nothing else { continue }
+
+                if newTypes[v] == nil {
+                    newTypes[v] = t
+                    numUpdatesPerVariable[v] = 1
+                } else {
+                    newTypes[v]! |= t
+                    numUpdatesPerVariable[v]! += 1
+                }
+            }
+        }
+
+        for (v, c) in numUpdatesPerVariable {
+            // Skip updating:
+            // 1. Fallback type for grandchild (no record in prevState)
+            // 2. Local variables, which go out of scope (prevState is .nothing)
+            if prevState[v] == nil || prevState[v] == .nothing {
+                newTypes.remove(v)
+                continue
+            }
+
+            // Not all paths updates this variable, so it must be unioned with the previous type
+            if c != states.count {
+                newTypes[v]! |= prevState[v]!
+            }
+        }
+
+        // Update currentState and add typeChanges
+        for (v, newType) in newTypes {
+            if currentState[v] != newType {
+                currentState[v] = newType
+                typeChanges.append((v, newType))
+            }
+        }
+    }
+
+    /// Return the type of the given variable.
+    /// Return .unknown for variables not available in this state.
+    public func type(of variable: Variable) -> Type {
+        return currentState[variable] ?? .unknown
+    }
+
+    // Set type change on stateChangeStack at given index
+    private mutating func setTypeChange(of v: Variable, to t: Type, at index: Int) {
+        stateChangesStack[index][stateChangesStack[index].count - 1][v] = t
+    }
+
+    /// Set the type of the given variable in the current state.
+    public mutating func setType(of v: Variable, to t: Type) {
+        // Save type in parent state if it is not already
+        if stateChangesStack[stateChangesStack.count - 2].last![v] == nil {
+            setTypeChange(of: v, to: currentState[v] ?? .nothing, at: stateChangesStack.count - 2)
```

Comment:
Same here, maybe `stateChangesStack.count - 2` could become some kind of enum value `parentState` or so?